### PR TITLE
Improve correctness witness visualization

### PR DESCRIPTION
**Pull request by vmordan**
Issue #59

### CODE REVIEW

File: bridge/reports/templates/reports/UnsafeContent.html [hunk 1]
```diff
@@ -41,6 +41,16 @@
                             <i class="ui sitemap icon"></i>
                         </button>
                     </div>
+                    <div class="ui buttons icon tiny orange">
```
Comment (vmordan) [inline bridge/reports/templates/reports/UnsafeContent.html]: Button should be green here

File: bridge/reports/static/reports/js/ETV.js [hunk 2]
```diff
@@ -600,6 +600,95 @@ $(document).ready(function () {
         $('.ETV_ShowCode').click();
         $(this).removeClass('disabled');
     });
+    var hidden_arr = [];
+    $('#ldv_button').click(function () {
```
Comment (vmordan) [inline bridge/reports/static/reports/js/ETV.js]: ldv_button - bad name

File: bridge/reports/static/reports/js/ETV.js [hunk 3]
```diff
@@ -600,6 +600,95 @@ $(document).ready(function () {
...
+                if (hidden_arr[i]) {
+                    children[i].style.display = "none";
+                } else {
+                    children[i].style.display = "inline";
+                }
+            }
+            $(this).removeClass('x');
+        } else {
+            $(this).addClass('x');
+            var variable = window.prompt("Enter variable name", "ldv_");
```
Comment (vmordan) [inline bridge/reports/static/reports/js/ETV.js]: remove prefix

File: bridge/reports/etv.py [hunk 4]
```diff
@@ -802,13 +1062,31 @@ def __get_source(self, file_name):
...
+                            if src[0] == '!':
+                                src = self.__parse_line(src)
+                                src = src[2:-1]
+                            elif ' == 0' in src and line.find(src) == -1:
+                                src = src.replace(' == 0', '')
+                                src = self.__parse_line(src)
+                            else:
+                                src = self.__parse_line(src)
+                            if '512' in src:
+                                src = '((int )prrs->pkt_flag & 512) != 0'
```
Comment (vmordan) [inline bridge/reports/etv.py]: what is this?

File: bridge/reports/etv.py [hunk 5]
```diff
@@ -21,6 +21,10 @@
 import tempfile
 import uuid
 import zipfile
+import pycparser
```
Comment (vmordan) [inline bridge/reports/etv.py]: update new modules in readme file - `pycparser sympy`

File: bridge/reports/etv.py [hunk 6]
```diff
@@ -169,6 +173,7 @@ def __init__(self, data, include_assumptions, thread_id, triangles, cnt=0):
     def add_line(self, edge, index):
         line = str(edge['start line']) if 'start line' in edge else None
         code = edge['source'] if 'source' in edge and len(edge['source']) > 0 else None
+
```
Comment (vmordan) [inline bridge/reports/etv.py]: Remove newline

File: bridge/reports/etv.py [hunk 7]
```diff
@@ -645,10 +764,14 @@ def __process_correctness_witness(self):
             self.data['edges'].append(start_edge)
             new_elem = dict(start_edge)
             del new_elem['enter']
+            tmp0 = 0
```
Comment (vmordan) [inline bridge/reports/etv.py]: strange variable name

File: bridge/reports/templates/reports/UnsafeContent.html [hunk 8]
```diff
@@ -41,6 +41,16 @@
...
+                         <button class="ui tiny button icon note-popup" id="ldv_button" data-content="{% trans 'Search for variables' %}">
+                             <i class="ui search icon"></i>
+                         </button>
+                    </div>
+                    <div class="ui buttons icon tiny orange">
+                         <button class="ui tiny button icon note-popup" id="false_cond_button" data-content="{% trans 'Show only red conditions' %}">
```
Comment (vmordan) [inline bridge/reports/templates/reports/UnsafeContent.html]: 'Show only red conditions' - confusing name

File: bridge/reports/templates/reports/UnsafeContent.html (hunk 1 above)
Comment (PavelAndrianov) [reply bridge/reports/templates/reports/UnsafeContent.html]: Could you attach an image? What is the button?

Comment (vmordan) [reply bridge/reports/templates/reports/UnsafeContent.html]: image
2 new buttons are orange and added in different sections

Comment (PavelAndrianov) [reply bridge/reports/templates/reports/UnsafeContent.html]: I agree, that the section should be one for two new buttons, but may be, different from old ones. But I am not sure, that the color should be the same.
Green buttons expand the trace? And new one is for search.

**Comment by vmordan** [comment]
image
Currently all conditions with the same line number are unified even if they correspond to different files. This should be fixed

**Comment by vmordan** [comment]
If you apply this branch to the second example from [correctness_witnesses.zip](https://github.com/mutilin/klever/files/3852070/correctness_witnesses.zip), the given invariant disappears

**Comment by vmordan** [comment]
As I noticed, the given changes increases wall time of processing correctness witnesses in about 5 times.

**Pull request merged**